helix-term/src/commands.rs

```diff
@@ -3896,6 +3901,105 @@ fn surround_delete(cx: &mut Context) {
     })
 }
 
+#[derive(Eq, PartialEq)]
+enum ShellBehavior {
+    Replace,
+    Insert,
+    Append,
+    Filter,
+    None,
+}
+
+fn shell_pipe(cx: &mut Context) {
+    shell(cx, "pipe:", true, ShellBehavior::Replace);
+}
+
+fn shell_pipe_ignore(cx: &mut Context) {
+    shell(cx, "pipe:", true, ShellBehavior::None);
+}
+
+fn shell_insert(cx: &mut Context) {
+    shell(cx, "command:", false, ShellBehavior::Insert);
+}
+
+fn shell_append(cx: &mut Context) {
+    shell(cx, "command:", false, ShellBehavior::Append);
+}
+
+fn shell_filter(cx: &mut Context) {
+    shell(cx, "filter:", true, ShellBehavior::Filter);
+}
+
+fn shell(cx: &mut Context, prompt: &str, pipe: bool, behavior: ShellBehavior) {
+    use std::io::Write;
+    use std::process::{Command, Stdio};
+    let shell = Option::<Vec<_>>::None; // this should come from config, but can't currently
+    let shell = match shell {
+        Some(v) if !v.is_empty() => v,
+        _ => {
+            if cfg!(windows) {
+                vec!["cmd".to_owned(), "/C".to_owned()]
+            } else {
+                vec!["sh".to_owned(), "-c".to_owned()]
+            }
+        }
+    };
+    let prompt = Prompt::new(
+        prompt.to_owned(),
+        Some('!'),
```

Comment:
I think we want to use the `|` register for all of these.

Reply:
OK